tests/lac/tensor_product_matrix_07.cc

```diff
@@ -0,0 +1,207 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+// Test TensorProductMatrixSymmetricSum for zero (constrained) rows and colums.
+
+#include <deal.II/dofs/dof_handler.h>
+
+#include <deal.II/fe/fe_dgq.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_q1.h>
+
+#include <deal.II/grid/grid_generator.h>
+
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/tensor_product_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include "../tests.h"
+
+#include "../testmatrix.h"
+
+template <int dim, typename Number>
+std::tuple<FullMatrix<Number>, FullMatrix<Number>>
+compute_reference_matrices(const unsigned int fe_degree)
+{
+  MappingQ1<dim> mapping;
+  QGauss<dim>    quadrature(fe_degree + 1);
+  FE_DGQ<dim>    fe(fe_degree);
+
+  Triangulation<dim> tria;
+  GridGenerator::hyper_cube(tria);
+
+  DoFHandler<dim> dof_handler(tria);
+  dof_handler.distribute_dofs(fe);
+
+  const unsigned int n_dofs_1D = fe.n_dofs_per_cell();
+
+  FullMatrix<Number> mass_matrix_reference(n_dofs_1D, n_dofs_1D);
+  FullMatrix<Number> derivative_matrix_reference(n_dofs_1D, n_dofs_1D);
+
+  FEValues<dim> fe_values(mapping,
+                          fe,
+                          quadrature,
+                          update_values | update_gradients | update_JxW_values);
+
+  fe_values.reinit(dof_handler.begin());
+
+  for (const unsigned int q_index : fe_values.quadrature_point_indices())
+    for (const unsigned int i : fe_values.dof_indices())
+      for (const unsigned int j : fe_values.dof_indices())
+        {
+          mass_matrix_reference(i, j) +=
+            (fe_values.shape_value(i, q_index) *
+             fe_values.shape_value(j, q_index) * fe_values.JxW(q_index));
+
+          derivative_matrix_reference(i, j) +=
+            (fe_values.shape_grad(i, q_index) *
+             fe_values.shape_grad(j, q_index) * fe_values.JxW(q_index));
+        }
+
+
+
+  return {mass_matrix_reference, derivative_matrix_reference};
+}
+
+template <typename Number>
+void
+print(const FullMatrix<Number> &matrix)
+{
+  matrix.print_formatted(deallog.get_file_stream(), 2, false, 7);
```

Comment:
Is any of the numbers you print sensitive to roundoff, say is any of the `0.13` printed below actually `0.125`? That would make the test sensitive to roundoff.

Reply:
I have switched to scientific notation and increased the precision.